multiple build failures with SDK 8.1 and 64-bit cmake

Building with VS2013 but more importantly SDK 8.1 as well as a recent
cmake, 3.7.2, results in several failures:

## doesn't find x64 dbghelp b/c cmake is now 64-bit

It's b/c cmake is now 64-bit:
```
% file /c/PROGRA~1/CMAKE/bin/cmake.exe
/c/PROGRA~1/CMAKE/bin/cmake.exe: PE32+ executable (console) x86-64, for MS Windows
```

64-bit config:
```
-- Looking for dbghelp here: |/bin/x64/dbghelp.dll;/tools/tracing/amd64/dbghelp.dll;C:\Program Files/Microsoft Visual Studio */Common7/IDE/Remote Debugger/x64/dbghelp.dll;C:\Program Files/Windows Kits/*/Debuggers/x64/dbghelp.dll;C:\Program Files/Microsoft Visual Studio */Common7/IDE/Remote Debugger/x64/dbghelp.dll;C:\Program Files/Debugging Tools for Windows (x64)/dbghelp.dll|
CMake Error at CMakeLists.txt:587 (message):
  dbghelp.dll required and not found
```

32-bit config:
```
-- Looking for dbghelp here: |/bin/x86/dbghelp.dll;/tools/tracing/i386/dbghelp.dll;C:\Program Files/Microsoft Visual Studio */Common7/IDE/Remote Debugger/x86/dbghelp.dll;C:\Program Files/Windows Kits/*/Debuggers/x86/dbghelp.dll;C:\Program Files/Microsoft Visual Studio */Common7/IDE/dbghelp.dll;C:\Program Files/Debugging Tools for Windows/dbghelp.dll|
CMake Error at CMakeLists.txt:587 (message):
  dbghelp.dll required and not found
```

```
% find /c/Program\ Files -iname dbghelp.dll
/c/Program Files/Microsoft SQL Server/110/Shared/dbghelp.dll
/c/Program Files/Windows Defender/DbgHelp.dll
find: '/c/Program Files/WindowsApps': Permission denied

08:53 PM ~/drmemory/git/build_x64_dbg
% find /c/Program\ Files\ \(x86\)/ -iname dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin/dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin/debug/dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin/release/dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin64/dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin64/debug/dbghelp.dll
/c/Program Files (x86)/Dr. Memory/bin64/release/dbghelp.dll
/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/dbghelp.dll
/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/Remote Debugger/x64/dbghelp.dll
/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/Remote Debugger/x86/dbghelp.dll
/c/Program Files (x86)/Microsoft Visual Studio 12.0/Team Tools/Performance Tools/dbghelp.dll
/c/Program Files (x86)/Microsoft Visual Studio 12.0/Team Tools/Performance Tools/x64/dbghelp.dll
/c/Program Files (x86)/Windows Kits/10/Debuggers/x64/dbghelp.dll
/c/Program Files (x86)/Windows Kits/10/Debuggers/x64/srcsrv/dbghelp.dll
/c/Program Files (x86)/Windows Kits/10/Debuggers/x86/dbghelp.dll
/c/Program Files (x86)/Windows Kits/10/Debuggers/x86/srcsrv/dbghelp.dll
/c/Program Files (x86)/Windows Kits/8.1/Debuggers/x64/dbghelp.dll

win10 cmd64:
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files

cmd32:
ProgramFiles=C:\Program Files (x86)
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
```

## app_suite/fibers_win.cpp fails to build

C:\derek\drmemory\git\src\tests\app_suite\fibers_win.cpp(73) : error C3861: 'FlsGetValue': identifier not found

need fibersapi.h on win8+

```
$ grep -r 'define VER_PRODUCTBUILD ' /extsw/win32/visual_studio/headers*
/extsw/win32/visual_studio/headers-SDK5.0/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   3790
/extsw/win32/visual_studio/headers-SDK5.0/Include/tpcver.h:#define VER_PRODUCTBUILD 3027
/extsw/win32/visual_studio/headers-SDK6.0a/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   6000
/extsw/win32/visual_studio/headers-SDK6.0a/Include/tpcver.h:#define VER_PRODUCTBUILD 3027
/extsw/win32/visual_studio/headers-SDK7.0/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   7600
/extsw/win32/visual_studio/headers-SDK7.0/Include/tpcver.h:#define VER_PRODUCTBUILD 3027
/extsw/win32/visual_studio/headers-SDK7.0a/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   7600
/extsw/win32/visual_studio/headers-SDK7.0a/Include/tpcver.h:#define VER_PRODUCTBUILD 3027
/extsw/win32/visual_studio/headers-SDK7.1/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   7600
/extsw/win32/visual_studio/headers-SDK7.1/Include/tpcver.h:#define VER_PRODUCTBUILD 3027
/extsw/win32/visual_studio/headers-SDK8.0/Include/shared/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   9200
/extsw/win32/visual_studio/headers-SDK8.1/Include/shared/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   9600
/extsw/win32/visual_studio/headers-VS2005/Include/ntverp.h:#define VER_PRODUCTBUILD            /* NT */   3790
```

=>
```
#include <ntverp.h>
#if VER_PRODUCTBUILD >= 9200
```

## registers asm "error A2"

```
Microsoft (R) Macro Assembler Version 12.00.21005.1
```

```
C:\derek\drmemory\git\build_x86_dbg> C:\PROGRA~2\MICROS~1.0\VC\bin\ml.exe  /nologo /Zi /Zd /c /Fotests\CMakeFiles\registers.dir\registers.c_asm.asm.obj tests\CMakeFiles\registers.dir\registers.c_asm.asm.obj.s"
 Assembling: tests\CMakeFiles\registers.dir\registers.c_asm.asm.obj.s
tests\CMakeFiles\registers.dir\registers.c_asm.asm.obj.s(1818
tests\CMakeFiles\registers.dir\registers.c_asm.asm.obj.s(1818) : error A2
```

```
1817:        mov      ecx, dword ptr [4 + eax]
1818:        lock xchg ecx, dword ptr [0 + edx]
1819:        mov      eax, dword ptr [0 + edx]
```

Removing the "lock" prefix makes it work.
It's redundant: an xchg that references memory automatically does a lock.
This is a bizarre error message though with no further text and with a
number that does not match the online ml.exe error numbers (fatal are
1000+, non-fatal are 2000+).

## build failures in namespace_tests_win.cpp

C:\derek\drmemory\git\src\tests\app_suite\namespace_tests_win.cpp(37) : error C3861: 'CreateBoundaryDescriptor': identifier not found
C:\derek\drmemory\git\src\tests\app_suite\namespace_tests_win.cpp(68) : error C3861: 'CreatePrivateNamespace': identifier not found
C:\derek\drmemory\git\src\tests\app_suite\namespace_tests_win.cpp(74) : error C3861: 'OpenPrivateNamespace': identifier not found

It's b/c we're using STRLESS instead of VERSION_LESS when setting
_WIN32_WINNT, and "10" is STRLESS than "6".

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -523,17 +523,27 @@ if (WIN32)
   # 5.2 does not work (it's not just slower w/o SymSearch: it fails).
   # Haven't tested in between.
   # WINDDK/3790.1830/bin/x86/dbghelp.dll is 6.3.
-  if (X64)
-    set(PROGFILES "$ENV{PROGRAMW6432}") # cmake is 32-bit
-    set(PROGFILES32 "$ENV{PROGRAMFILES}")
-    set(ARCH_SFX "x64")
-    set(DDK_SFX "amd64")
-  else (X64)
+  # Older CMake binaries are 32-bit but newer ones can be 64-bit so we
+  # cannot rely on one or the other.
+  if ("$ENV{PROGRAMW6432}" STREQUAL "")
+    if (X64)
+      message(FATAL_ERROR "On 32-bit Windows: 64-bit build not supported")
+    endif ()
     set(PROGFILES "$ENV{PROGRAMFILES}")
     set(PROGFILES32 "$ENV{PROGRAMFILES}")
     set(ARCH_SFX "x86")
     set(DDK_SFX "i386")
-  endif (X64)
+  else ()
+    set(PROGFILES "$ENV{PROGRAMW6432}")
+    set(PROGFILES32 "$ENV{PROGRAMFILES(x86)}")
+    if (X64)
+      set(ARCH_SFX "x64")
+      set(DDK_SFX "amd64")
+    else (X64)
+      set(ARCH_SFX "x86")
+      set(DDK_SFX "i386")
+    endif (X64)
+  endif ()
   # Even the VS2005 copy is 6.5 (despite its headers being < 6.3) so we can
   # use those as well as the later SDK and standalone DTFW versions.
   set(dbghelp_paths

--- a/make/policies.cmake
+++ b/make/policies.cmake
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2015 Google, Inc.  All rights reserved.
+# Copyright (c) 2015-2017 Google, Inc.  All rights reserved.
 # **********************************************************
 
 # Dr. Memory: the memory debugger
@@ -20,6 +20,7 @@
 
 if ("${CMAKE_VERSION}" VERSION_EQUAL "3.1" OR
     "${CMAKE_VERSION}" VERSION_GREATER "3.1")
+  cmake_policy(SET CMP0053 OLD)
   cmake_policy(SET CMP0054 OLD)
 endif ()
 

--- a/tests/app_suite/CMakeLists.txt
+++ b/tests/app_suite/CMakeLists.txt
@@ -63,11 +63,11 @@ elseif (WIN32)
   # i#1283 set _WIN32_WINNT for app_suite power_tests_win test
   # XXX: we did not do every-version-set to avoid over-specifying, however,
   # we could hit similar problems of different win-versions in the future
-  if ("${CMAKE_SYSTEM_VERSION}" STRLESS "6.1")
+  if ("${CMAKE_SYSTEM_VERSION}" VERSION_LESS "6.1")
     # pre-win7, set _WIN32_WINNT to be _WIN32_WINNT_WINXP
     append_compile_flags(app_suite_tests "-D_WIN32_WINNT=0x0501")
   else ()
-    # pre-win7, set _WIN32_WINNT to be _WIN32_WINNT_WIN7
+    # win7+: set _WIN32_WINNT to be _WIN32_WINNT_WIN7
     append_compile_flags(app_suite_tests "-D_WIN32_WINNT=0x0601")
   endif ()
   if (NOT CMAKE_C_COMPILER_VERSION VERSION_LESS 18.0)

--- a/tests/app_suite/fibers_win.cpp
+++ b/tests/app_suite/fibers_win.cpp
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2012-2014 Google, Inc.  All rights reserved.
+ * Copyright (c) 2012-2017 Google, Inc.  All rights reserved.
  * Copyright (c) 2004 VMware, Inc.  All rights reserved.
  * **********************************************************/
 
@@ -52,6 +52,10 @@
 
 #include "gtest/gtest.h"
 #include <windows.h>
+#include <ntverp.h>
+#if VER_PRODUCTBUILD >= 9200 /* win8+ SDK */
+# include "fibersapi.h"
+#endif
 #include "stdio.h"
 
 static DWORD flsA_index;

--- a/tests/registers.c
+++ b/tests/registers.c
@@ -1069,7 +1069,8 @@ GLOBAL_LABEL(FUNCNAME:)
          * sub-dword, so we have to read 4 bytes and then expand the bottom 2 bytes.
          */
         mov      ecx, DWORD [4 + REG_XAX]
-        lock xchg ecx, DWORD [0 + REG_XDX] /* undef => slowpath */
+        /* "lock" is automatic here, and adding it can cause ml to fail (i#1952) */
+        xchg ecx, DWORD [0 + REG_XDX] /* undef => slowpath */
         mov      eax, DWORD [0 + REG_XDX] /* propagate lock==0xf0 unless xl8 cleared */
         movzx    eax, ax /* now 0x00==defined unless xl8 cleared */
         cmp      eax, edx /* undef w/ proper clearing */

--- a/tests/registers.res
+++ b/tests/registers.res
@@ -115,7 +115,7 @@ registers.c_asm.asm:1773
 Error #25: UNINITIALIZED READ: reading 2 byte(s)
 registers.c_asm.asm:1787
 Error #26: UNINITIALIZED READ: reading register eax
-registers.c_asm.asm:1821
+registers.c_asm.asm:1822
 %endif
 %if UNIX
 Error #22: UNINITIALIZED READ: reading 1 byte(s)
@@ -127,7 +127,7 @@ registers.c_asm.asm:1092
 Error #25: UNINITIALIZED READ: reading 2 byte(s)
 registers.c_asm.asm:1106
 Error #26: UNINITIALIZED READ: reading register eax
-registers.c_asm.asm:1075
+registers.c_asm.asm:1076
 %endif
 %OUT_OF_ORDER
 : LEAK 15 direct bytes + 0 indirect bytes

